File: src/tests/Behavioral/UnsafePointerWordRead/UnsafePointerWordRead.go
package main

import (
	"fmt"
	"unsafe"
)

// The two arms of the pointer-word read `*(*unsafe.Pointer)(unsafe.Pointer(&x))`:
//
//   - a uintptr source's word IS the number, and must round-trip exactly;
//   - a reference-kinded source (a channel here) yields a NON-NIL word for a live value —
//     the nil-bit is the contract time.syncTimer's consumer reads.
//
// The un-fixed emission punned the source's first reference-sized bits into an
// unsafe.Pointer REFERENCE: junk dispatch on a quiet heap, an AccessViolationException
// when the bits landed unmapped (the asynctimerchan=2 witness, 2026-08-24).
func main() {
	u := uintptr(0xC0FFEE)
	q := *(*unsafe.Pointer)(unsafe.Pointer(&u))
	fmt.Println("uintptr word round-trips:", uintptr(q) == 0xC0FFEE)

	c := make(chan int, 1)
	p := *(*unsafe.Pointer)(unsafe.Pointer(&c))
	fmt.Println("live channel word is non-nil:", p != nil)

	// The word is STABLE for the same storage: two reads agree.
	p2 := *(*unsafe.Pointer)(unsafe.Pointer(&c))
	fmt.Println("re-read agrees:", uintptr(p) == uintptr(p2))
}

File: src/tests/Behavioral/UnsafePointerWordRead/UnsafePointerWordRead.cs
namespace go;

using fmt = fmt_package;
using @unsafe = unsafe_package;

partial class main_package {

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object uintptrWordRoundTripsˢ = (@string)"uintptr word round-trips:"u8;
private static readonly object liveChannelWordIsNonNilˢ = (@string)"live channel word is non-nil:"u8;
private static readonly object reReadAgreesˢ = (@string)"re-read agrees:"u8;

internal static void Main() {
    ref var u = ref heap<uintptr>(out var Ꮡu);
    u = (uintptr)0xC0FFEE;
    @unsafe.Pointer q = ~Ꮡ(new @unsafe.Pointer(~Ꮡu));
    fmt.Println(uintptrWordRoundTripsˢ, (uintptr)q == 0xC0FFEE);
    ref var c = ref heap<channel<nint>>(out var Ꮡc);
    c = new channel<nint>(1);
    @unsafe.Pointer p = ~Ꮡ(new @unsafe.Pointer((uintptr)Ꮡc));
    fmt.Println(liveChannelWordIsNonNilˢ, p != nil);
    @unsafe.Pointer p2 = ~Ꮡ(new @unsafe.Pointer((uintptr)Ꮡc));
    fmt.Println(reReadAgreesˢ, (uintptr)p == (uintptr)p2);
}

} // end main_package
